cumulusci/tests/pytest_plugins/pytest_sf_orgconnect.py

```diff
@@ -85,36 +143,123 @@ def create_task(task_class, options=None, project_config=None, org_config=None):
-    test_dir = request.node.fspath.dirname
-    if is_integration_test:
-        return os.path.join(test_dir, "large_cassettes")  # 8-tracks
-    else:  # standard behaviour from
-        # https://github.com/ktosiek/pytest-vcr/blob/master/pytest_vcr.py
-        return os.path.join(test_dir, "cassettes")
+def classify_and_modify_test(item, marker_names):
+    if "slow" in marker_names:
+        if not item.config.getoption("--run-slow-tests"):
+            pytest.skip("slow: test requires --run-slow-tests")
+
+    if "large_vcr" in marker_names:
+        library_dir = Path(cumulusci.__file__).parent.parent / "large_cassettes"
+        item.add_marker(pytest.mark.vcr(cassette_library_dir=str(library_dir)))
+
+        test_name = vcr_cassette_name_for_item(item)
+        test_path = Path(library_dir) / (test_name + ".yaml")
+        if not (test_path.exists() or item.config.getoption("--org")):
+            pytest.skip("large_vcr: test requires --org to generate large cassette")
+
+    elif "needs_org" in marker_names and not item.config.getoption("--org"):
+        pytest.skip("needs_org: test requires --org")
+
+
+# a cache of orgs matching shapes:
+#    TODO test that it is properly used
+org_shapes = ContextVar("org_shapes", default={})
+
+
+@pytest.fixture(
+    scope="function",
+)
+def run_code_without_recording(request, vcr, user_requested_network_access):
+    def really_run_code_without_recording(func):
+        if user_requested_network_access:
+            # Run the setup code, but don't record it
+            with set_mode(RECORDING, False), set_mode(READING, False):
+                return func()
+
+    return really_run_code_without_recording
+
+
+class CurrentOrg:
+    __slots__ = ("org_config",)
+
+    def __init__(self):
+        self.org_config = None
+
+
+@pytest.fixture(scope="session")
+def current_org_shape(request):
+    """Internal: shared state for fixtures"""
+    org = CurrentOrg()
+    return org
+
+
+@pytest.fixture(scope="function", autouse=True)
+def _change_org_shape(request, current_org_shape):
+    """Select an org_shape for a test
+    e.g.:
+
+        @pytest.mark.org_shape("qa", "qa_org")
+        def test_foo(create_task);
+            t = create_task(Foo)
+            t()
+
+    Switch the current org to an org created with
+    org template "qa" after running flow "qa_org".
+
+    Clean up any changes you make, because this org may be reused by
+    other tests.
+    """
+    marker = request.node.get_closest_marker("org_shape")
+    if marker:
+        config_name, flow_name = marker.args
+        with change_org_shape(current_org_shape, config_name, flow_name) as org_config:
+            yield org_config
+    else:
+        yield None
+
+
+@contextmanager
+def change_org_shape(current_org_shape, config_name: str, flow_name: str):
+    # I don't love that we're using the user's real keychain
+    # but I get weird popups when I use an empty keychain.
```

Comment:
Do we leak the org after the test suite runs? I don't see anything that would remove it.

Reply:
There's a limit on active scratch orgs in a dev hub. It could be problematic if we get to the point where a run of CCI's integration tests creates a lot of scratch orgs and doesn't clean them up.